source-bigquery-batch/driver.go

```diff
@@ -0,0 +1,698 @@
+	// discarded it, in order to keep the `discoverTables` / `discoverPrimaryKeys`
+	// interface as stupidly simple as possible, which ought to make it just that
+	// little bit easier to do this generically for multiple databases.
+	var keysByTable = make(map[string]*discoveredPrimaryKey)
+	for _, key := range keys {
+		keysByTable[key.Schema+"."+key.Table] = key
+	}
+
+	var bindings []*pc.Response_Discovered_Binding
+	for _, table := range tables {
+		// Exclude tables in "system schemas" such as information_schema or pg_catalog.
+		if slices.Contains(drv.ExcludedSystemSchemas, table.Schema) {
+			continue
+		}
+
+		var tableID = table.Schema + "." + table.Name
+
+		var recommendedName = recommendedCatalogName(table.Name)
+		var res, err = drv.GenerateResource(recommendedName, table.Schema, table.Name, table.Type)
+		if err != nil {
+			log.WithFields(log.Fields{
+				"reason": err,
+				"table":  tableID,
+				"type":   table.Type,
+			}).Warn("unable to generate resource spec")
+			continue
+		}
+		resourceConfigJSON, err := json.Marshal(res)
+		if err != nil {
+			return nil, fmt.Errorf("error serializing resource spec: %w", err)
+		}
+
+		// Try to generate a useful collection schema, but on error fall back to the
+		// minimal schema with the default key [/_meta/polled, /_meta/index].
+		var collectionSchema = minimalSchema
+		var collectionKey = fallbackKey
+		if tableKey, ok := keysByTable[tableID]; ok {
+			if generatedSchema, err := generateCollectionSchema(tableKey.Columns, tableKey.ColumnTypes); err == nil {
+				collectionSchema = generatedSchema
+				collectionKey = nil
+				for _, colName := range tableKey.Columns {
+					collectionKey = append(collectionKey, primaryKeyToCollectionKey(colName))
+				}
+			} else {
+				log.WithFields(log.Fields{"table": tableID, "err": err}).Warn("unable to generate collection schema")
+			}
+		}
+
+		bindings = append(bindings, &pc.Response_Discovered_Binding{
+			RecommendedName:    recommendedName,
+			ResourceConfigJson: resourceConfigJSON,
+			DocumentSchemaJson: collectionSchema,
+			Key:                collectionKey,
+		})
+	}
+
+	return &pc.Response_Discovered{Bindings: bindings}, nil
+}
+
+// primaryKeyToCollectionKey converts a database primary key column name into a Flow collection key
+// JSON pointer with escaping for '~' and '/' applied per RFC6901.
+func primaryKeyToCollectionKey(key string) string {
+	// Any encoded '~' must be escaped first to prevent a second escape on escaped '/' values as
+	// '~1'.
+	key = strings.ReplaceAll(key, "~", "~0")
+	key = strings.ReplaceAll(key, "/", "~1")
+	return "/" + key
+}
+
+const queryDiscoverTables = `SELECT table_schema, table_name, table_type FROM %[1]s.INFORMATION_SCHEMA.TABLES;`
+
+type discoveredTable struct {
+	Schema string
+	Name   string
+	Type   string // Usually 'BASE TABLE' or 'VIEW'
+}
+
+func discoverTables(ctx context.Context, db *bigquery.Client, dataset string) ([]*discoveredTable, error) {
```

Comment:
Oh, that would work. I was aware that it could be done by executing more SQL queries to list them all, but that seemed overly complicated and so I just did the simplest thing that could work for now. I'm not sure how people use multiple datasets in BigQuery, but it seemed like common use-cases probably wouldn't span multiple datasets anyway, so I'm also inclined to just leave it like this for now.